pkg/controller/common.go

```diff
@@ -1100,26 +1101,57 @@ func verifyObjectFailedOrDeleted(obj interface{}, objectDeleted bool) (bool, err
 	return false, nil
 }
 
-// WatchObject creates a noop controller to set up a watch and an informer for an object in controller-runtime manager
-// Use this function if you want to set up a watch for an object without a configuring a separate informer factory.
-func WatchObject(mgr manager.Manager, objKind source.Kind) error {
-	objType := objKind.Type.GetName()
-	c, err := controller.New(fmt.Sprintf("watch %s", objType), mgr, controller.Options{
+// WatchNamespace creates a noop controller to set up a watch and an informer for Namespace objects in controller-runtime manager
+func CacheNamespace(mgr manager.Manager) error {
+	c, err := controller.New("namespaces-controller", mgr, controller.Options{
 		Reconciler: &noOpReconciler{},
 	})
 	if err != nil {
 		return err
 	}
 
-	c.GetLogger().Info("Starting to watch %s", objType)
+	c.GetLogger().Info("Starting to watch namespaces")
 
-	// Watch for CRUD events on azVolumeAttachment objects
-	err = c.Watch(&objKind, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+	// Watch for CRUD events on Namespace objects
+	err = c.Watch(&source.Kind{Type: &v1.Namespace{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
 		CreateFunc: func(_ event.CreateEvent) bool {
 			return false
 		},
 		UpdateFunc: func(_ event.UpdateEvent) bool {
-			return true
+			return false
+		},
+		GenericFunc: func(_ event.GenericEvent) bool {
+			return false
+		},
+		DeleteFunc: func(_ event.DeleteEvent) bool {
+			return false
+		},
+	})
+	return err
+}
+
+// WatchCRD creates a noop controller to set up a watch and an informer for CRD objects in controller-runtime manager
+func CacheCRD(mgr manager.Manager, sharedState *SharedState) error {
+	c, err := controller.New("CRD-controller", mgr, controller.Options{
+		Reconciler: &noOpReconciler{},
+	})
+	if err != nil {
+		return err
+	}
+
+	c.GetLogger().Info("Starting to watch CRDs")
+
+	// Watch for CRUD events on CRD objects
+	err = c.Watch(&source.Kind{Type: &apiext.CustomResourceDefinition{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		CreateFunc: func(_ event.CreateEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			crdObj, ok := e.ObjectNew.(*apiext.CustomResourceDefinition)
+			if ok && (crdObj.Name == consts.AzVolumeAttachmentCRDName || crdObj.Name == consts.AzVolumeCRDName) && crdObj.DeletionTimestamp != nil {
+				atomic.StoreUint32(&sharedState.driverUninstall, 1)
```

Comment:
We don't want to block the main thread so if we were going to handle things in a predicate, it would need to be done in a goroutine. And we probably should only update `if !sharedState.IsDriverUninstall()`